File: src/main/java/leet_code/OddEvenLinkedList.go
package main

import "fmt"

func main() {

}

type ListNode struct {
	Val  int
	Next *ListNode
}

func oddEvenList(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}

	if head.Next == nil {
		return head
	}
	current := head

	// [2,1,3,5,6,4,7]

	odd := current
	even := current.Next
	current = even.Next
	evenHead := even

	for current != nil {
		odd.Next = current // take odd one
		odd = current

		even.Next = current.Next // take even one
		even = current.Next

		fmt.Println(even)
		fmt.Println(odd)

		if even == nil {
			break
		}
		current = current.Next.Next
	}

	odd.Next = evenHead

	return head
}

File: src/main/java/leet_code/OddEvenLinkedList.java
package leet_code;

public class OddEvenLinkedList {


    public static void main(String[] args) {
        //[1,2,3,4,5]

        var head = new ListNode(1);

        var current = new ListNode(2);
        head.next = current;

        current.next = new ListNode(3);


    }


    public static class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }

        @Override
        public String toString() {
            return " " + val + " ";
        }
    }

    public ListNode oddEvenList(ListNode head) {
        throw new IllegalArgumentException("Not impl yet");
    }

}
